fuel-merkle/src/sparse/merkle_tree.rs

```diff
@@ -144,14 +141,15 @@ where
         Ok(())
     }
 
-    pub fn delete(&mut self, key: &Bytes32) -> Result<(), MerkleTreeError<StorageError>> {
+    pub fn delete(&mut self, key: &Bytes32, data: &[u8]) -> Result<(), MerkleTreeError<StorageError>> {
         if self.root() == *Self::empty_root() {
             // The zero root signifies that all leaves are empty, including the
             // given key.
             return Ok(());
         }
 
-        if let Some(primitive) = self.storage.get(key)? {
+        let leaf_node = Node::create_leaf(key, data);
+        if let Some(primitive) = self.storage.get(&leaf_node.hash())? {
```

Comment:
Actually, I checked the code, and it seems you don't even need a value. Because `self.path_set`  and `self.delete_with_path_set` uses only `leaf_key` inside the logic.

Reply:
Does this mean we wouldn't need to pass the data here?

Reply:
Yeap

Reply:
Interesting, I can explore that further

Reply:
Unfortunately, this doesn't work: We can't avoid the storage lookup here because we need to know if the deleted item is in the set of leaves or not. Whenever we call delete, we will modify all items along the leaf path. If we never added the item, we end up "reverting" a leaf path that was never actually changed, thus creating a false root. I got far enough into the changes in a PR here https://github.com/FuelLabs/fuel-vm/pull/483, but the tests with delete are failing for this reason.

We need to be able to short-circuit the delete call before modifying the nodes along the leaf path. As far as I am aware, the only way to do this is by checking the storage.

Reply:
@xgreenx observed that we can check for the leaf by querying the path set, allowing us to properly check the delete 👍